File: python/zebra-puzzle/zebra_puzzle.py
from itertools import permutations


_nations = ["Norwegian", "Ukrainian", "Englishman", "Spaniard", "Japanese"]
yellow, blue, red, ivory, green = range(5)
fox, horse, snails, dog, zebra = range(5)
water, tea, milk, orange_juice, coffee = range(5)
kools, chesterfield, old_gold, lucky_strike, parliament = range(5)


def drinks_water():
    for house in solution():
        if house["drink"] == water:
            return house["nation"]


def owns_zebra():
    for house in solution():
        if house["pet"] == zebra:
            return house["nation"]


def solution():
    for nations in permutations(_nations):
        for colors in permutations(range(5)):
            for pets in permutations(range(5)):
                for drinks in permutations(range(5)):
                    for smokes in permutations(range(5)):
                        houses = [{
                            "nation": nations[i],
                            "color": colors[i],
                            "pet": pets[i],
                            "drink": drinks[i],
                            "smoke": smokes[i],
                        } for i in range(5)]

                        if all(func(houses) for func in check_funcs):
                            return houses


def known1(houses):
    return len(houses) == 5


def known2(houses):
    return any(house["nation"] == "Englishman"
               and house["color"] == red for house in houses)


def known3(houses):
    return any(house["nation"] == "Spaniard"
               and house["pet"] == dog for house in houses)


def known4(houses):
    return any(house["drink"] == coffee
               and house["color"] == green for house in houses)


def known5(houses):
    return any(house["nation"] == "Ukrainian"
               and house["drink"] == tea for house in houses)


def known6(houses):
    return any(houses[i + 1]["color"] == green
               and houses[i]["color"] == ivory for i in range(4))


def known7(houses):
    return any(house["smoke"] == old_gold
               and house["pet"] == snails for house in houses)


def known8(houses):
    return any(house["smoke"] == kools
               and house["color"] == yellow for house in houses)


def known9(houses):
    return houses[2]["drink"] == milk


def known10(houses):
    return houses[0]["nation"] == "Norwegian"


def known11(houses):
    for i in range(5):
        if houses[i]["smoke"] == chesterfield:
            if i > 0 and houses[i - 1]["pet"] == fox:
                return True
            if i < 4 and houses[i + 1]["pet"] == fox:
                return True
            return False


def known12(houses):
    for i in range(5):
        if houses[i]["smoke"] == kools:
            if i > 0 and houses[i - 1]["pet"] == horse:
                return True
            if i < 4 and houses[i + 1]["pet"] == horse:
                return True
            return False


def known13(houses):
    return any(house["smoke"] == lucky_strike
               and house["drink"] == orange_juice for house in houses)


def known14(houses):
    return any(house["nation"] == "Japanese"
               and house["smoke"] == parliament for house in houses)


def known15(houses):
    for i in range(5):
        if houses[i]["nation"] == "Norwegian":
            if i > 0 and houses[i - 1]["color"] == blue:
                return True
            if i < 4 and houses[i + 1]["color"] == blue:
                return True
            return False


check_funcs = [
    known1,
    known2,
    known3,
    known4,
    known5,
    known6,
    known7,
    known8,
    known9,
    known10,
    known11,
    known12,
    known13,
    known14,
    known15,
]

File: haskell/zebra-puzzle/src/ZebraPuzzle.hs
module ZebraPuzzle
  ( Resident(..)
  , Solution(..)
  , solve
  ) where

import           Data.List                      ( find
                                                , permutations
                                                , zip5
                                                )
import           Data.Maybe                     ( fromJust )

data Color = Yellow | Blue | Red | Ivory | Green
  deriving (Bounded, Enum, Eq)

data Pet = Fox | Horse | Snail | Dog | Zebra
  deriving (Bounded, Enum, Eq)

data Drink = Water | Tea | Milk | OrangeJuice | Coffee
  deriving (Bounded, Enum, Eq)

data Cigarette = Kools | Chesterfield | OldGold | LuckyStrike | Parliament
  deriving (Bounded, Enum, Eq)

data Resident = Englishman | Spaniard | Ukrainian | Norwegian | Japanese
  deriving (Bounded, Enum, Eq, Show)

data Solution = Solution
  { waterDrinker :: Resident
  , zebraOwner   :: Resident
  }
  deriving (Eq, Show)

solve :: Solution
solve = Solution (findWater result) (findZebra result)

findWater :: [(Color, Pet, Drink, Cigarette, Resident)] -> Resident
findWater ((_, _, Water, _, resident) : _) = resident
findWater (_ : xs) = findWater xs
findWater _ = error ""

findZebra :: [(Color, Pet, Drink, Cigarette, Resident)] -> Resident
findZebra ((_, Zebra, _, _, resident) : _) = resident
findZebra (_ : xs) = findZebra xs
findZebra _ = error ""

result :: [(Color, Pet, Drink, Cigarette, Resident)]
result = fromJust (find checkConditions xs)
 where
  xs =
    [ zip5 a b c d e
    | a <- permutations ([minBound .. maxBound] :: [Color])
    , b <- permutations ([minBound .. maxBound] :: [Pet])
    , c <- permutations ([minBound .. maxBound] :: [Drink])
    , d <- permutations ([minBound .. maxBound] :: [Cigarette])
    , e <- permutations ([minBound .. maxBound] :: [Resident])
    ]

checkConditions :: [(Color, Pet, Drink, Cigarette, Resident)] -> Bool
checkConditions xs = all (\f -> f xs) conditions

conditions :: [[(Color, Pet, Drink, Cigarette, Resident)] -> Bool]
conditions =
  [ (== 5) . length
  , any (\(col, _, _, _, res) -> col == Red && res == Englishman)
  , any (\(_, pet, _, _, res) -> pet == Dog && res == Spaniard)
  , any (\(col, _, dri, _, _) -> col == Green && dri == Coffee)
  , any (\(_, _, dri, _, res) -> dri == Tea && res == Ukrainian)
  , condition6
  , any (\(_, pet, _, cig, _) -> pet == Snail && cig == OldGold)
  , any (\(col, _, _, cig, _) -> col == Yellow && cig == Kools)
  , \(_ : _ : (_, _, dri, _, _) : _) -> dri == Milk
  , \((_, _, _, _, res) : _) -> res == Norwegian
  , condition11
  , condition12
  , any (\(_, _, dri, cig, _) -> dri == OrangeJuice && cig == LuckyStrike)
  , any (\(_, _, _, cig, res) -> cig == Parliament && res == Japanese)
  , condition15
  ]
 where
  condition6 ((Ivory, _, _, _, _) : (Green, _, _, _, _) : _) = True
  condition6 (_ : xs) = condition6 xs
  condition6 _        = False
  condition11 ((_, Fox, _, _, _) : (_, _, _, Chesterfield, _) : _) = True
  condition11 ((_, _, _, Chesterfield, _) : (_, Fox, _, _, _) : _) = True
  condition11 (_ : xs) = condition11 xs
  condition11 _        = False
  condition12 ((_, Horse, _, _, _) : (_, _, _, Kools, _) : _) = True
  condition12 ((_, _, _, Kools, _) : (_, Horse, _, _, _) : _) = True
  condition12 (_ : xs) = condition12 xs
  condition12 _        = False
  condition15 ((Blue, _, _, _, _) : (_, _, _, _, Norwegian) : _) = True
  condition15 ((_, _, _, _, Norwegian) : (Blue, _, _, _, _) : _) = True
  condition15 (_ : xs) = condition11 xs
  condition15 _        = False
